plugins/xdebug/xdebug.cc

```diff
@@ -242,6 +243,63 @@ InjectMilestonesHeader(TSHttpTxn txn, TSMBuffer buffer, TSMLoc hdr)
   }
 }
 
+static const char NotFound[] = "Not-Found";
+
+// The returned string must be freed with TSfree() unless it is equal to the constant NotFound.
+//
+static const char *
+getRemapUrlStr(TSHttpTxn txnp, TSReturnCode (*remapUrlGetFunc)(TSHttpTxn, TSMLoc *), int &urlStrLen)
+{
+  TSMLoc urlLoc;
+
+  TSReturnCode rc = remapUrlGetFunc(txnp, &urlLoc);
+  if (rc != TS_SUCCESS) {
+    urlStrLen = sizeof(NotFound) - 1;
+    return NotFound;
+  }
+
+  char *urlStr = TSUrlStringGet(nullptr, urlLoc, &urlStrLen);
+
+  // Be defensive.
+  if ((urlStrLen == 0) and urlStr) {
+    TSError("[xdebug] non-null remap URL string with zero length");
+    TSfree(urlStr);
+    urlStr = nullptr;
+  }
+
+  if (!urlStr) {
+    urlStrLen = sizeof(NotFound) - 1;
+    return NotFound;
+  }
+
+  return urlStr;
+}
+
+static void
+InjectRemapHeader(TSHttpTxn txn, TSMBuffer buffer, TSMLoc hdr)
+{
+  TSMLoc dst = FindOrMakeHdrField(buffer, hdr, "X-Remap", lengthof("X-Remap"));
+
+  if (TS_NULL_MLOC != dst) {
+    int fromUrlStrLen, toUrlStrLen;
+    const char *fromUrlStr = getRemapUrlStr(txn, TSRemapFromUrlGet, fromUrlStrLen);
+    const char *toUrlStr   = getRemapUrlStr(txn, TSRemapToUrlGet, toUrlStrLen);
+
+    char buf[2048];
+    int len = snprintf(buf, sizeof(buf) - 1, "from=%*s, to=%*s", fromUrlStrLen, fromUrlStr, toUrlStrLen, toUrlStr);
```

Comment:
Should be sizeof(buf):
     The snprintf() and vsnprintf() functions will write at most size-1 of the characters printed into the output string (the size'th
     character then gets the terminating `\0'); if the return value is greater than or equal to the size argument, the string was too
     short and some of the printed characters were discarded.  The output is always null-terminated, unless size is 0.

Reply:
That could be considered a separate issue from this PR, as I followed the pattern I saw in the existing function InjectTxnUuidHeader() .